doc/man/doveadm-quota.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-QUOTA 1 "2015-05-09" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-quota \- Initialize/recalculate or show current quota usage
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.BR quota " [" \-S
.IR socket_path "] " command
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.BR quota " [" \-S
.IR socket_path "]"
.IB command\  \-A
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.BR quota " [" \-S
.IR socket_path "]"
.IB command\  \-F \ file
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] [" \-f
.IR formatter ]
.BR quota " [" \-S
.IR socket_path "]"
.IB  command\  \-u \ user
.\"------------------------------------------------------------------------
.SH DESCRIPTION
In the first form,
.BR doveadm (1)
will execute the
.I command
for the currently logged in user.
.PP
In the second form, the
.I command
will be performed for all users.
.PP
In the third form, the
.I command
will be performed for all users listed in the given
.IR file .
.PP
In the fourth form, the
.I command
will affect only the matching
.IR user (s).
.IP Note:
The
.B quota get
and
.B quota recalc
commands are only available when the global
.I mail_plugins
setting contains the
.B quota
plugin.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\" --- command specific options --- "/.
.PP
Command specific
.IR options :
.\"-------------------------------------
@INCLUDE:option-A@
.\"-------------------------------------
@INCLUDE:option-F-file@
.\"-------------------------------------
@INCLUDE:option-S-socket@
.\"-------------------------------------
@INCLUDE:option-u-user@
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS quota get
.BR doveadm " [" \-f
.IR formatter ]
.B quota get
[\fB\-A\fP|\fB\-u\fP \fIuser\fP|\fB\-F\fP \fIfile\fP]
.PP
The
.B quota get
command is used to display the current quota usage.
The storage values are reported in kilobytes.
.PP
This command uses by default the output formatter
.BR table .
.\"-------------------------------------
.SS quota recalc
.B doveadm quota recalc
[\fB\-A\fP|\fB\-u\fP \fIuser\fP|\fB\-F\fP \fIfile\fP]
.PP
The
.B quota recalc
command is used to recalculate the current quota usage.
.\"------------------------------------------------------------------------
.SH FILES
.TP
.I @pkgsysconfdir@/dovecot.conf
Dovecot\(aqs main configuration file, including the
.I dict
section.
.TP
.I @pkgsysconfdir@/dovecot\-dict\-sql.conf.ext
SQL dictionary proxy settings.
.TP
.I @pkgsysconfdir@/conf.d/10\-mail.conf
Mailbox locations and namespaces, including global
.I mail_location
.TP
.I @pkgsysconfdir@/conf.d/90\-quota.conf
Quota configuration.
.\"------------------------------------------------------------------------
.SH EXAMPLE
Get the current quota usage of user jane.
.sp
.nf
.B doveadm quota get \-u jane
Quota name                        Type    Value  Limit  %
user                              STORAGE 90099 102400 87
user                              MESSAGE 20548  30000 68
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.1 2018-02-29  Aki Tuomi <aki.tuomi@dovecot.fi>

	* Submission server support improvements and bug fixes
	  - Lots of bug fixes to submission server
	* API CHANGE: array_idx_modifiable will no longer allocate space
	  - Particularly affects how you should check MODULE_CONTEXT result, or
	    use REQUIRE_MODULE_CONTEXT.

	+ mail_attachment_detection_options setting controls when
	  $HasAttachment and $HasNoAttachment keywords are set for mails.
	+ imap: Support fetching body snippets using FETCH (SNIPPET) or
	  (SNIPPET (LAZY=FUZZY))
	+ fs-compress: Automatically detect whether input is compressed or not.
	  Prefix the compression algorithm with "maybe-" to enable the
	  detection, for example: "compress:maybe-gz:6:..."
	+ Added settings to change dovecot.index* files' optimization behavior.
	  See https://wiki2.dovecot.org/IndexFiles#Settings
	+ Auth cache can now utilize auth workers to do password hash
	  verification by setting auth_cache_verify_password_with_worker=yes.
	+ Added charset_alias plugin. See
	  https://wiki2.dovecot.org/Plugins/CharsetAlias
	+ imap_logout_format and pop3_logout_format settings now support all of
	  the generic variables (e.g. %{rip}, %{session}, etc.)
	+ Added auth_policy_check_before_auth, auth_policy_check_after_auth
	  and auth_policy_report_after_auth settings.
	+ master: Support HAProxy PP2_TYPE_SSL command and set "secured"
	  variable appropriately
	- Invalid UCS4 escape in HTML can cause crashes
	- imap: IMAP COMPRESS -enabled clietn crashes on disconnect
	- lmtp: Fix crash when user is over quota
	- lib-lda: Parsing Return-Path header address fails when it contains
	  CFWS
	- auth: SASL with Exim fails for AUTH commands without an initial
	  response
	- imap: SPECIAL-USE capability isn't automatically added
	- auth: LDAP subqueries do not support standard auth variables in
	  var-expand
	- auth: SHA256-CRYPT and SHA512-CRYPT schemes do not work
	- lib-index: mail_always/never_cache_fields are not used for existing
	  cache files
	- imap: Fetching headers leaks memory if search doesn't find any mails
	- lmtp: ORCPT support in RCPT TO
	- imap-login: Process sometimes ends up in infinite loop
	- sdbox: Rolled back save/copy transaction doesn't delete temp files
	- mail: lock_method=dotlock causes crashes

v2.3.0.1 2018-02-28  Timo Sirainen <tss@iki.fi>

	* CVE-2017-15130: TLS SNI config lookups may lead to excessive
	  memory usage, causing imap-login/pop3-login VSZ limit to be reached
	  and the process restarted. This happens only if Dovecot config has
	  local_name { } or local { } configuration blocks and attacker uses
	  randomly generated SNI servernames.
	* CVE-2017-14461: Parsing invalid email addresses may cause a crash or
	  leak memory contents to attacker. For example, these memory contents
	  might contain parts of an email from another user if the same imap
	  process is reused for multiple users. First discovered by Aleksandar
	  Nikolic of Cisco Talos. Independently also discovered by "flxflndy"
	  via HackerOne.
	* CVE-2017-15132: Aborted SASL authentication leaks memory in login
	  process.
	* Linux: Core dumping is no longer enabled by default via
	  PR_SET_DUMPABLE, because this may allow attackers to bypass
	  chroot/group restrictions. Found by cPanel Security Team. Nowadays
	  core dumps can be safely enabled by using "sysctl -w
	  fs.suid_dumpable=2". If the old behaviour is wanted, it can still be
	  enabled by setting:
	  import_environment=$import_environment PR_SET_DUMPABLE=1
	- imap-login with SSL/TLS connections may end up in infinite loop

v2.3.0 2017-12-22  Timo Sirainen <tss@iki.fi>

	* Various setting changes, see https://wiki2.dovecot.org/Upgrading/2.3
	* Logging rewrite started: Logging is now based on hierarchical events.
	  This makes it possible to do various things, like: 1) giving
	  consistent log prefixes, 2) enabling debug logging with finer
	  granularity, 3) provide logs in more machine readable formats
	  (e.g. json). Everything isn't finished yet, especially a lot of the
	  old logging code still needs to be translated to the new way.
	* Statistics rewrite started: Stats are now based on (log) events.
	  It's possible to gather statistics about any event that is logged.
	  See http://wiki2.dovecot.org/Statistics for details
	* ssl_dh setting replaces the old generated ssl-parameters.dat
	* IMAP: When BINARY FETCH finds a broken mails, send [PARSE] error
	  instead of [UNKNOWNCTE]
	* Linux: core dumping via PR_SET_DUMPABLE is no longer enabled by
	  default due to potential security reasons (found by cPanel Security
	  Team).

	+ Added support for SMTP submission proxy server, which includes
	  support for BURL and CHUNKING extension.
	+ LMTP rewrite. Supports now CHUNKING extension and mixing of
	  local/proxy recipients.
	+ auth: Support libsodium to add support for ARGON2I and ARGON2ID
	  password schemes.
	+ auth: Support BLF-CRYPT password scheme in all platforms
	+ auth: Added LUA scripting support for passdb/userdb.
	  See https://wiki2.dovecot.org/AuthDatabase/Lua
	- Input streams are more reliable now when there are errors or when
	  the maximum buffer size is reached. Previously in some situations
	  this could have caused Dovecot to try to read already freed memory.
	- Output streams weren't previously handling failures when writing a
	  trailer at the end of the stream. This mainly affected encrypt and
	  zlib compress ostreams, which could have silently written truncated
	  files if the last write happened to fail (which shouldn't normally
	  have ever happened).
	- virtual plugin: Fixed panic when fetching mails from virtual
	  mailboxes with IMAP BINARY extension.
	- doveadm-server: Fix potential hangs with SSL connections
	- doveadm proxy: Reading commands' output from v2.2.33+ servers could
	  have caused the output to be corrupted or caused a crash.
	- Many other smaller fixes

v2.2.33.2 2017-10-20  Timo Sirainen <tss@iki.fi>


src/plugins/quota/doveadm-quota.c
/* Copyright (c) 2005-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "module-dir.h"
#include "quota-plugin.h"
#include "quota-private.h"
#include "doveadm-print.h"
#include "doveadm-mail.h"

const char *doveadm_quota_plugin_version = DOVECOT_ABI_VERSION;

void doveadm_quota_plugin_init(struct module *module);
void doveadm_quota_plugin_deinit(void);

static void cmd_quota_get_root(struct quota_root *root)
{
	const char *const *res;
	uint64_t value, limit;
	int ret;

	res = quota_root_get_resources(root);
	for (; *res != NULL; res++) {
		ret = quota_get_resource(root, "", *res, &value, &limit);
		doveadm_print(root->set->name);
		doveadm_print(*res);
		if (ret > 0) {
			doveadm_print_num(value);
			doveadm_print_num(limit);
			if (limit > 0)
				doveadm_print_num(value*100 / limit);
			else
				doveadm_print("0");
		} else if (ret == 0) {
			doveadm_print_num(value);
			doveadm_print("-");
			doveadm_print("0");
		} else {
			doveadm_print("error");
			doveadm_print("error");
			doveadm_print("error");
		}
	}
}

static int
cmd_quota_get_run(struct doveadm_mail_cmd_context *ctx ATTR_UNUSED,
		  struct mail_user *user)
{
	struct quota_user *quser = QUOTA_USER_CONTEXT(user);
	struct quota_root *const *root;

	if (quser == NULL) {
		i_error("Quota not enabled");
		doveadm_mail_failed_error(ctx, MAIL_ERROR_NOTFOUND);
		return -1;
	}

	array_foreach(&quser->quota->roots, root)
		cmd_quota_get_root(*root);
	return 0;
}

static void cmd_quota_get_init(struct doveadm_mail_cmd_context *ctx ATTR_UNUSED,
			       const char *const args[] ATTR_UNUSED)
{
	doveadm_print_header("root", "Quota name", 0);
	doveadm_print_header("type", "Type", 0);
	doveadm_print_header("value", "Value",
			     DOVEADM_PRINT_HEADER_FLAG_RIGHT_JUSTIFY);
	doveadm_print_header("limit", "Limit",
			     DOVEADM_PRINT_HEADER_FLAG_RIGHT_JUSTIFY);
	doveadm_print_header("percent", "%",
			     DOVEADM_PRINT_HEADER_FLAG_RIGHT_JUSTIFY);
}

static struct doveadm_mail_cmd_context *
cmd_quota_get_alloc(void)
{
	struct doveadm_mail_cmd_context *ctx;

	ctx = doveadm_mail_cmd_alloc(struct doveadm_mail_cmd_context);
	ctx->v.run = cmd_quota_get_run;
	ctx->v.init = cmd_quota_get_init;
	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	return ctx;
}

static int
cmd_quota_recalc_run(struct doveadm_mail_cmd_context *ctx ATTR_UNUSED,
		     struct mail_user *user)
{
	struct quota_user *quser = QUOTA_USER_CONTEXT(user);
	struct quota_root *const *root;
	struct quota_transaction_context trans;

	if (quser == NULL) {
		i_error("Quota not enabled");
		doveadm_mail_failed_error(ctx, MAIL_ERROR_NOTFOUND);
		return -1;
	}

	i_zero(&trans);
	trans.quota = quser->quota;
	trans.recalculate = QUOTA_RECALCULATE_FORCED;

	array_foreach(&quser->quota->roots, root) {
		(void)(*root)->backend.v.update(*root, &trans);
		if ((*root)->backend.v.flush != NULL)
			(*root)->backend.v.flush(*root);
	}
	return 0;
}

static struct doveadm_mail_cmd_context *
cmd_quota_recalc_alloc(void)
{
	struct doveadm_mail_cmd_context *ctx;

	ctx = doveadm_mail_cmd_alloc(struct doveadm_mail_cmd_context);
	ctx->v.run = cmd_quota_recalc_run;
	return ctx;
}

static struct doveadm_cmd_ver2 quota_commands[] = {
	{
		.name = "quota get",
		.usage = "",
		.mail_cmd = cmd_quota_get_alloc,
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAMS_END
	},
	{
		.name = "quota recalc",
		.usage = "",
		.mail_cmd = cmd_quota_recalc_alloc,
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_MAIL_COMMON
DOVEADM_CMD_PARAMS_END
	}
};

void doveadm_quota_plugin_init(struct module *module ATTR_UNUSED)
{
	unsigned int i;

	for (i = 0; i < N_ELEMENTS(quota_commands); i++)
		doveadm_cmd_register_ver2(&quota_commands[i]);
}

void doveadm_quota_plugin_deinit(void)
{
}
